assets/js/hocs/with-store-cart-api-hydration.js

```diff
@@ -13,13 +13,50 @@ import { useSelect } from '@wordpress/data';
  */
 const useStoreCartApiHydration = () => {
 	const cartData = useRef( getSetting( 'cartData' ) );
+	const { cartDataIsStale } = useDispatch( CART_STORE_KEY );
 
 	useSelect( ( select, registry ) => {
 		if ( ! cartData.current ) {
 			return;
 		}
+		const { isResolving, hasFinishedResolution, isCartDataStale } = select(
+			CART_STORE_KEY
+		);
 
-		const { isResolving, hasFinishedResolution } = select( CART_STORE_KEY );
+		/**
+		 * This should only execute once. When the code further down the file executes
+		 * then the condition directly below this comment should never evaluate to true
+		 * on subsequent executions. Because of this localStorage.getItem won't be
+		 * called multiple times.
+		 */
+		if (
+			! isCartDataStale() &&
+			! isResolving( 'getCartData' ) &&
+			! hasFinishedResolution( 'getCartData', [] )
+		) {
+			const lastCartUpdateRaw = window.localStorage.getItem(
+				'lastCartUpdate'
+			);
+			const lastCartUpdate =
+				lastCartUpdateRaw === null
+					? null
+					: JSON.parse( lastCartUpdateRaw );
+
+			if ( lastCartUpdate?.timestamp === undefined ) {
+				cartDataIsStale( false );
+				return;
+			}
+
+			const needsUpdateFromAPI =
+				lastCartUpdate.timestamp >
+				cartData.current?.generated_timestamp;
+
+			if ( needsUpdateFromAPI ) {
+				cartDataIsStale();
+				return;
+			}
+			cartDataIsStale( false );
```

Comment:
I guess this line is needed because there is no default value set under metadata for cartDataIsStale. Would it be an issue to start with cartData.metaData.cartDataIsStale: false?